File: BackEnd/model/StudyGroup.js
import mongoose from 'mongoose';

const MemberSchema = new mongoose.Schema({
    user_id: { 
        type: mongoose.Schema.Types.ObjectId, 
        ref: 'User', 
        required: true 
    },
    role: { 
        type: String, 
        enum: ['member', 'admin'], 
        default: 'member' 
    },
    joined_at: { 
        type: Date, 
        default: Date.now 
    }
}, { _id: false });

const StudyGroupSchema = new mongoose.Schema({
    name: { 
        type: String, 
        required: true, 
        trim: true,
        index: true 
    },
    
    description: { 
        type: String, 
        trim: true 
    },
    
    avatar: { 
        type: String,
        default: 'https://via.placeholder.com/150' 
    },
    
    level: { 
        type: String, 
        enum: ['N5', 'N4', 'N3', 'N2', 'N1', 'ALL'], 
        default: 'ALL',
        index: true 
    },
    
    creator_id: { 
        type: mongoose.Schema.Types.ObjectId, 
        ref: 'User', 
        required: true,
        index: true 
    },
    
    members: {
        type: [MemberSchema],
        default: []
    },
    
    member_count: {
        type: Number,
        default: 0,
        min: 0
    },
    
    is_active: {
        type: Boolean,
        default: true,
        index: true
    },
    
    is_private: {
        type: Boolean,
        default: false
    },
    
    max_members: {
        type: Number,
        default: 50,
        min: 1
    }

}, { 
    timestamps: true 
});

// tìm kiếm nhanh theo thành viên, tên nhóm, mô tả, creator và trạng thái
StudyGroupSchema.index({ 'members.user_id': 1 });
StudyGroupSchema.index({ name: 'text', description: 'text' });
StudyGroupSchema.index({ creator_id: 1, is_active: 1 });

// lấy tổng số thành viên
StudyGroupSchema.virtual('total_members').get(function() {
    return this.members.length;
});

// kiểm tra user có phải thành viên không
StudyGroupSchema.methods.isMember = function(userId) {
    return this.members.some(m => m.user_id.toString() === userId.toString());
};

// kiểm tra user có phải admin không
StudyGroupSchema.methods.isAdmin = function(userId) {
    return this.members.some(
        m => m.user_id.toString() === userId.toString() && m.role === 'admin'
    );
};

// kiểm tra user có phải creator không
StudyGroupSchema.methods.isCreator = function(userId) {
    return this.creator_id.toString() === userId.toString();
};

// thêm thành viên
StudyGroupSchema.methods.addMember = function(userId, role = 'member') {
    if (this.isMember(userId)) {
        throw new Error('Người dùng đã là thành viên của nhóm');
    }
    
    if (this.members.length >= this.max_members) {
        throw new Error('Nhóm đã đạt số lượng thành viên tối đa');
    }
    
    this.members.push({
        user_id: userId,
        role: role,
        joined_at: new Date()
    });
    
    this.member_count = this.members.length;
};

// xóa thành viên
StudyGroupSchema.methods.removeMember = function(userId) {
    const initialLength = this.members.length;
    this.members = this.members.filter(
        m => m.user_id.toString() !== userId.toString()
    );
    
    if (this.members.length === initialLength) {
        throw new Error('Người dùng không phải là thành viên của nhóm');
    }
    
    this.member_count = this.members.length;
};

export default mongoose.model('StudyGroup', StudyGroupSchema);

File: FrontEnd/lib/features/study_groups/models/study_group.dart
import '../../../core/network/api_client.dart';

class StudyGroupMember {
  final String userId;
  final String role;
  final DateTime joinedAt;
  final String? fullName;
  final String? username;
  final String? avatar;
  final String? email;

  StudyGroupMember({
    required this.userId,
    required this.role,
    required this.joinedAt,
    this.fullName,
    this.username,
    this.avatar,
    this.email,
  });

  factory StudyGroupMember.fromJson(Map<String, dynamic> json) {
    final userData = json['user'] ?? json['user_id'];
    String? fullName;
    String? username;
    String? email;
    String? avatarField;

    if (userData is Map) {
      fullName = userData['full_name'] ??
          userData['display_name'] ??
          userData['name'] ??
          userData['HoTen'];
      username = userData['username'] ??
          userData['user_name'] ??
          userData['TenDangNhap'];
      email = userData['email'] ?? userData['Email'];
      avatarField = ApiClient.resolveAssetUrl(
        userData['AnhDaiDien'] ?? userData['avatar'],
      );
    }

    return StudyGroupMember(
      userId: userData is Map ? userData['_id'] : userData.toString(),
      role: json['role'] ?? 'member',
      joinedAt:
          DateTime.parse(json['joined_at'] ?? DateTime.now().toIso8601String()),
      fullName: fullName,
      username: username,
      avatar: avatarField,
      email: email,
    );
  }

  bool get isAdmin => role == 'admin';
}

class StudyGroup {
  final String id;
  final String name;
  final String? description;
  final String? avatar;
  final String level;
  final String creatorId;
  final List<StudyGroupMember> members;
  final int memberCount;
  final bool isActive;
  final bool isPrivate;
  final int maxMembers;
  final DateTime createdAt;
  final DateTime updatedAt;

  // Creator info (nếu được populate)
  final String? creatorName;
  final String? creatorAvatar;
  final String? creatorUsername;

  StudyGroup({
    required this.id,
    required this.name,
    this.description,
    this.avatar,
    required this.level,
    required this.creatorId,
    required this.members,
    required this.memberCount,
    required this.isActive,
    required this.isPrivate,
    required this.maxMembers,
    required this.createdAt,
    required this.updatedAt,
    this.creatorName,
    this.creatorAvatar,
    this.creatorUsername,
  });

  factory StudyGroup.fromJson(Map<String, dynamic> json) {
    final creatorData = json['creator_id'];

    final avatarUrl = ApiClient.resolveAssetUrl(json['avatar']);

    return StudyGroup(
      id: json['_id'] ?? json['id'],
      name: json['name'] ?? '',
      description: json['description'],
      avatar: avatarUrl ?? 'https://via.placeholder.com/150',
      level: json['level'] ?? 'ALL',
      creatorId:
          creatorData is Map ? creatorData['_id'] : creatorData.toString(),
      members: (json['members'] as List?)
              ?.map((m) => StudyGroupMember.fromJson(m))
              .toList() ??
          [],
      memberCount: json['member_count'] ?? 0,
      isActive: json['is_active'] ?? true,
      isPrivate: json['is_private'] ?? false,
      maxMembers: json['max_members'] ?? 50,
      createdAt:
          DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt:
          DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
      creatorName: creatorData is Map
          ? creatorData['full_name'] ??
              creatorData['display_name'] ??
              creatorData['HoTen']
          : null,
      creatorAvatar: creatorData is Map
          ? ApiClient.resolveAssetUrl(
              creatorData['AnhDaiDien'] ?? creatorData['avatar'],
            )
          : null,
      creatorUsername: creatorData is Map
          ? creatorData['username'] ?? creatorData['TenDangNhap']
          : null,
    );
  }

  bool isFull() => memberCount >= maxMembers;

  bool isMember(String userId) {
    return members.any((m) => m.userId == userId);
  }

  bool isAdmin(String userId) {
    return members.any((m) => m.userId == userId && m.isAdmin);
  }

  bool isCreator(String userId) {
    return creatorId == userId;
  }

  String get levelDisplay {
    switch (level) {
      case 'ALL':
        return 'Tất cả cấp độ';
      case 'N1':
      case 'N2':
      case 'N3':
      case 'N4':
      case 'N5':
        return 'JLPT $level';
      default:
        return level;
    }
  }

  String get memberCountDisplay => '$memberCount/$maxMembers thành viên';
}

class GroupStats {
  final int totalMessages;
  final int todayMessages;
  final int activeMembers;
  final Map<String, int> messagesByUser;

  GroupStats({
    required this.totalMessages,
    required this.todayMessages,
    required this.activeMembers,
    required this.messagesByUser,
  });

  factory GroupStats.fromJson(Map<String, dynamic> json) {
    return GroupStats(
      totalMessages: json['total_messages'] ?? 0,
      todayMessages: json['today_messages'] ?? 0,
      activeMembers: json['active_members'] ?? 0,
      messagesByUser: Map<String, int>.from(json['messages_by_user'] ?? {}),
    );
  }
}
